### DISCUSSION TITLE
Serializing using toObject ignores populated fields

### DISCUSSION THREAD

**Discussion by xmajzel** [mikro-orm/mikro-orm#5968] [General]
### Describe the bug

When using **serialize([entityB], { populate: ['entitiesC'] })** in test bellow, I expect that nested **entityC** will have relations serialized as primaryKey references.

Instead it is serialized as I used **serialize([entityB], { populate: ['entitiesC.entityB.entityA'] })**.

It only happens, when I use **toObject** in custom serializer method.

### Reproduction

```typescript
import {
  MikroORM,
  Entity,
  Property,
  OneToMany,
  Collection,
  BaseEntity,
  BigIntType,
  PrimaryKey,
  ManyToOne,
  type Ref, serialize,
} from '@mikro-orm/postgresql';

@Entity({ abstract: true })
abstract class CustomBaseEntity extends BaseEntity {

  @PrimaryKey({ autoincrement: true, type: new BigIntType('string') })
  readonly id!: string;

}

@Entity()
class EntityC extends CustomBaseEntity {

  @ManyToOne(() => EntityB, { ref: true })
  entityB!: Ref<EntityB>;

  @Property()
  name!: string;

}


@Entity()
class EntityB extends CustomBaseEntity {

  @Property()
  name!: string;

  @ManyToOne(() => EntityA, { ref: true })
  entityA!: Ref<EntityA>;

  @OneToMany(() => EntityC, a => a.entityB, {
    serializer: (value: Collection<EntityC>) => {
      if (!value.isInitialized()) {
        return undefined;
      }
      return value
      .getItems()
      .map(a => a.toObject());
    },
  })
  entitiesC = new Collection<EntityC>(this);

}

@Entity()
class EntityA extends CustomBaseEntity {

  @OneToMany(() => EntityB, slot => slot.entityA)
  entitiesB = new Collection<EntityB>(this);

  @Property()
  name!: string;

}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [CustomBaseEntity, EntityC, EntityB, EntityA],
    dbName: 'gh-xxxx',
  });
  await orm.schema.dropSchema();
  await orm.schema.createSchema();
});


afterAll(async () => {
  await orm.schema.dropSchema();
  await orm.close(true);
});


afterEach(async () => {
  await orm.schema.clearDatabase();
});


test('Serialize newly created entity', async () => {
  let entityA = orm.em.create(EntityA, { name: 'I am entity A' });
  await orm.em.flush();
  orm.em.clear();

  entityA = await orm.em.findOneOrFail(EntityA, { id: entityA.id });

  const entityB = orm.em.create(EntityB, { entityA, name: 'I am entity B' });
  orm.em.create(EntityC, { entityB, name: 'I am entity C (1)' });
  orm.em.create(EntityC, { entityB, name: 'I am entity C (2)' });

  await orm.em.flush();

  const serialized = serialize([entityB], { populate: ['entitiesC'] });

  expect(serialized).toStrictEqual([
    {
      id: '1',
      entitiesC: [
        { id: '1', entityB: '1', name: 'I am entity C (1)' },
        { id: '2', entityB: '1', name: 'I am entity C (2)' },
      ],
      name: 'I am entity B',
      entityA: '1',
    },
  ]);
});
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.6

### Node.js version

20.11.1

### Operating system

Mac OS Sonoma 14.4.1 (23E224)

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
The combination of explicit serialization with `toObject` feels weird on its own, I am not entirely sure what you want from that code, but `toObject` will always work based on the populate hints (and it's your custom serializer function that is using that). 

You can disable custom serializers via `ignoreSerializers: true` option, or use explicit serialization inside your custom serializer (so `serialize()` instead of `toObject()`).

**Comment by xmajzel** [comment]
I will try to explain briefly, I want to achieve.

I want to use explicit serialization as `const serialized = serialize([entityB], { populate: ['entitiesC'] });`.
In custom serializer I want to keep serializationContext from parent (**so it will behave as I didn't have customSerializer**) and only filter some entities based on some condition, e.g.:
```diff
serializer: (value: Collection<EntityC>) => {
  if (!value.isInitialized()) {
    return undefined;
  }
  return value
  .getItems()
+ .filter(a => a.name !== 'Some value')
  .map(a => a.toObject());
}
```
So I want to last line from custom serializer (`.map(a => a.toObject());`) to take into account current serialization context.

Looking into **EntitySerializer.processCollection**, when serializing it's extracting childOptions for entity (`const childOptions = this.extractChildOptions(options, prop) as Dictionary;`), so it would be great if such options could be extracted and passed into custom serializer next to the **value** parameter.

Then I would be able to achieve it like follows:

```diff
-serializer: (value: Collection<EntityC>) => {
+serializer: (value: Collection<EntityC>, options: SerializeOptions) => {
  if (!value.isInitialized()) {
    return undefined;
  }
  return value
  .getItems()
  .filter(a => a.name !== 'Some value')
+ .map(a => serialize(a, options));
}
```

**Reply by B4nan** [reply]
Here you go, sorry for the long wait.

dbfddceeeadeb55a95c362b79f4983f28bbb0488

**Reply by xmajzel** [reply]
Thanks!
